### PR TITLE
[FEATURE][TOOLS-CDC]:add flink Command line support

### LINKED ISSUE #101: [Feature]:  Add command line tools to support Flink Jdbc Connector as source
### Describe your use case

Need a command line tool to start a job to migrate data to OceanBase using Flink Jdbc Connector.

### Describe the solution you'd like

Construct the options of Flink Jdbc Connector in the tool classes, and submit a job to sink data to OceanBase from the jdbc source.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xxsc0529**
## Summary

When synchronizing data to OceanBase, the flink CLI is supported. close #101

## Solution Description

Define program entries, receive parameters, and parse them. Use Oceanbase Catalog to obtain the information required by Flink and create the corresponding sink。Here's how it works.
`CREATE TABLE test_history (
    itemid BIGINT NOT NULL,
    clock INT DEFAULT 0 NOT NULL,
    value DOUBLE PRECISION DEFAULT 0.0000 NOT NULL,
    ns INT DEFAULT 0 NOT NULL,
    PRIMARY KEY (itemid, clock, ns)
) ;

CREATE TABLE test_history_uint (
    itemid BIGINT NOT NULL,
    clock INT DEFAULT 0 NOT NULL,
    value DECIMAL(20,0) DEFAULT 0 NOT NULL,
    ns INT DEFAULT 0 NOT NULL,
    PRIMARY KEY (itemid, clock, ns)
) ;

CREATE TABLE test_history_str (
    itemid BIGINT NOT NULL,
    clock INT DEFAULT 0 NOT NULL,
    value VARCHAR(255) DEFAULT '' NOT NULL,
    ns INT DEFAULT 0 NOT NULL,
    PRIMARY KEY (itemid, clock, ns)
);`
The command line is as follows
`./bin/flink run \
    -Dexecution.checkpointing.interval=10s \
    -Dparallelism.default=1 \
    -c com.oceanbase.connector.flink.tools.cdc.CdcTools \
    lib/flink-connector-oceanbase-tools-cdc-1.4-SNAPSHOT.jar \
    mysql-sync-database \
    --database test_db \
    --mysql-conf hostname=xxxx \
    --mysql-conf port=3306 \
    --mysql-conf username=root \
    --mysql-conf password=xxxx \
    --mysql-conf database-name=test_db \
    --including-tables "tbl1|test.*" \
    --sink-conf username=xxxx@xxxx \
    --sink-conf password=xxxx \
    --sink-conf url=jdbc:mysql://xxxx:xxxx  \
    --sink-conf  sink.label-prefix=label  \
    --table-conf  replication_num=1`
The sink-config configuration is the configuration of the OceanBase database

### CODE REVIEW

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSchemaFactory.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        }
+        sb.append(")");
+
+        // Append table comment
+        if (schema.getTableComment() != null && !schema.getTableComment().trim().isEmpty()) {
+            sb.append(" COMMENT='").append(quoteComment(schema.getTableComment())).append("'");
+        }
+        sb.append(";");
+
+        System.out.println("Generated DDL: " + sb);
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSchemaFactory.java:110]: Please use logger but not System.out, check it in other files as well.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/table/OceanBaseJsonSerializationSchema.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            Object[] values = new Object[size];
+            for (int i = 0; i < size; i++) {
+                String fieldName = tableInfo.getFieldNames().get(i);
+                JsonNode fieldNode = rowDataNode.path("after").path(fieldName);
+                values[i] = objectMapper.convertValue(fieldNode, new TypeReference<Object>() {});
+            }
+
+            return new DataChangeRecord(tableInfo, type, values);
+        } catch (IOException e) {
+            log.error("Failed to parse rowData JSON: {}", rowDataStr, e);
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/table/OceanBaseJsonSerializationSchema.java:74]: Why not throw the exception out?

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/FieldSchema.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.oceanbase.connector.flink.tools.catalog;
+
+public class FieldSchema {
+    private String name;
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/FieldSchema.java:20]: In what scenarios will these fields be modified? How about make these fields final and remove all setters?

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSchemaFactory.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        }
+
+        if (field.getComment() != null && !field.getComment().trim().isEmpty()) {
+            sb.append(" COMMENT '").append(quoteComment(field.getComment())).append("'");
+        }
+
+        sb.append(", ");
+    }
+
+    public static String identifier(String name) {
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSchemaFactory.java:134]: Better to use OceanBaseDialect.quoteIdentifier

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/TableSchema.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+package com.oceanbase.connector.flink.tools.catalog;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TableSchema {
+    public static final String DORIS_TABLE_REGEX = "^[a-zA-Z][a-zA-Z0-9-_]*$";
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/TableSchema.java:25]: Unrelated field.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/TableSchema.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    private String database;
+    private String table;
+    private String tableComment;
+    private Map<String, FieldSchema> fields;
+    private List<String> keys = new ArrayList<>();
+    private List<String> distributeKeys = new ArrayList<>();
+    private Map<String, String> properties = new HashMap<>();
+
+    private Integer tableBuckets;
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/TableSchema.java:34]: What is tableBuckets used for?

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/cdc/SourceSchema.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        this.databaseName = databaseName;
+        this.schemaName = schemaName;
+        this.tableName = tableName;
+        this.tableComment = tableComment;
+    }
+
+    public abstract String convertToOceanBaseType(
+            String fieldType, Integer precision, Integer scale);
+
+    public abstract String getCdcTableName();
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/cdc/SourceSchema.java:51]: Seems it is not necessary while there is already a method getTableIdentifier here.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/cdc/CdcTools.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        String tablePrefix = params.get(DatabaseSyncConfig.TABLE_PREFIX);
+        String tableSuffix = params.get(DatabaseSyncConfig.TABLE_SUFFIX);
+        String includingTables = params.get(DatabaseSyncConfig.INCLUDING_TABLES);
+        String excludingTables = params.get(DatabaseSyncConfig.EXCLUDING_TABLES);
+        String multiToOneOrigin = params.get(DatabaseSyncConfig.MULTI_TO_ONE_ORIGIN);
+        String multiToOneTarget = params.get(DatabaseSyncConfig.MULTI_TO_ONE_TARGET);
+        boolean createTableOnly = params.has(DatabaseSyncConfig.CREATE_TABLE_ONLY);
+        boolean ignoreDefaultValue = params.has(DatabaseSyncConfig.IGNORE_DEFAULT_VALUE);
+        boolean ignoreIncompatible = params.has(DatabaseSyncConfig.IGNORE_INCOMPATIBLE);
+        boolean singleSink = params.has(DatabaseSyncConfig.SINGLE_SINK);
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/cdc/CdcTools.java:131]: What is it used for?

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/cdc/DatabaseSyncConfig.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.oceanbase.connector.flink.tools.cdc;
+
+public class DatabaseSyncConfig {
+
+    public static final String MYSQL_SYNC_DATABASE = "mysql-sync-database";
```
Comment (whhe) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/cdc/DatabaseSyncConfig.java:21]: There should be a short description for every config key.

**Comment by whhe** [comment]
There should be at least a test case for mysql cdc, please add one on the github workflow.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/table/OceanBaseJsonSerializationSchema.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+import java.sql.Time;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class OceanBaseJsonSerializationSchema extends AbstractRecordSerializationSchema<String> {
+
+    private static final long serialVersionUID = 1L;
+    private static final Logger log =
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/table/OceanBaseJsonSerializationSchema.java:41]: It is best to use uppercase letters to be consistent with other places.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSchemaFactory.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        tableSchema.setDatabase(database);
+        tableSchema.setTable(table);
+        tableSchema.setFields(columnFields);
+        tableSchema.setKeys(buildKeys(pkKeys, columnFields));
+        tableSchema.setTableComment(tableComment);
+        tableSchema.setDistributeKeys(buildDistributeKeys(pkKeys, columnFields));
+        return tableSchema;
+    }
+
+    private static List<String> buildDistributeKeys(
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSchemaFactory.java:53]: What does DistributeKeys mean in oceanbase, oceanbase only has index keys and partition keys.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSchemaFactory.java [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        Map<String, FieldSchema> fields = schema.getFields();
+        List<String> keys = schema.getKeys();
+
+        // Append fields
+        for (Map.Entry<String, FieldSchema> entry : fields.entrySet()) {
+            FieldSchema field = entry.getValue();
+            buildColumn(sb, field, keys.contains(entry.getKey()));
+        }
+
+        sb = sb.deleteCharAt(sb.length() - 1); // 删除最后一个逗号
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSchemaFactory.java:92]: Do not use Chinese.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSystem.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+import java.sql.DatabaseMetaData;
+import java.sql.ResultSet;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** OceanBase System Operate. */
+@Public
+public class OceanBaseSystem implements Serializable {
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseSystem.java:44]: Better to change the name, it's confusing.

**Review by yuanoOo — CHANGES_REQUESTED** [review]
Thanks for your contribution, I left some comments.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String SMALLINT = "SMALLINT";
+    public static final String INT = "INT";
+    public static final String BIGINT = "BIGINT";
+    public static final String LARGEINT = "LARGEINT";
+    // largeint is bigint unsigned in information_schema.COLUMNS
+    public static final String BIGINT_UNSIGNED = "BIGINT UNSIGNED";
+    public static final String FLOAT = "FLOAT";
+    public static final String DOUBLE = "DOUBLE";
+    public static final String DECIMAL = "DECIMAL";
+    public static final String DECIMAL_V3 = "DECIMALV3";
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java:31]: What does this mean? ob does not support this type.

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+public class OceanBaseType {
+    public static final String BOOLEAN = "BOOLEAN";
+    public static final String TINYINT = "TINYINT";
+    public static final String SMALLINT = "SMALLINT";
+    public static final String INT = "INT";
+    public static final String BIGINT = "BIGINT";
+    public static final String LARGEINT = "LARGEINT";
+    // largeint is bigint unsigned in information_schema.COLUMNS
+    public static final String BIGINT_UNSIGNED = "BIGINT UNSIGNED";
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java:27]: Does ob support this type?

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String FLOAT = "FLOAT";
+    public static final String DOUBLE = "DOUBLE";
+    public static final String DECIMAL = "DECIMAL";
+    public static final String DECIMAL_V3 = "DECIMALV3";
+    public static final String DATE = "DATE";
+    public static final String DATE_V2 = "DATEV2";
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java:33]: And this

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String DATETIME = "DATETIME";
+    public static final String DATETIME_V2 = "DATETIMEV2";
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java:35]: And this

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    public static final String CHAR = "CHAR";
+    public static final String VARCHAR = "VARCHAR";
+    public static final String STRING = "STRING";
+    public static final String HLL = "HLL";
+    public static final String BITMAP = "BITMAP";
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseType.java:40]: And these three

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseTypeMapper.java [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+import static com.alibaba.druid.sql.ast.SQLDataType.Constants.BOOLEAN;
+import static com.oceanbase.connector.flink.tools.catalog.OceanBaseType.*;
+
+public class OceanBaseTypeMapper {
+
+    /** Max size of char type of OceanBase. */
+    public static final int MAX_CHAR_SIZE = 255;
+
+    /** Max size of varchar type of OceanBase. */
+    public static final int MAX_VARCHAR_SIZE = 65533;
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseTypeMapper.java:56]: I remember it was 262144

File: flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseTypeMapper.java [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+
+import java.sql.Types;
+
+import static com.alibaba.druid.sql.ast.SQLDataType.Constants.BOOLEAN;
+import static com.oceanbase.connector.flink.tools.catalog.OceanBaseType.*;
+
+public class OceanBaseTypeMapper {
+
+    /** Max size of char type of OceanBase. */
+    public static final int MAX_CHAR_SIZE = 255;
```
Comment (yuanoOo) [inline flink-connector-oceanbase-tools-cdc/src/main/java/com/oceanbase/connector/flink/tools/catalog/OceanBaseTypeMapper.java:53]: I remember it was 256

**Pull request closed without merging**